plugin/server_audit/server_audit.cc

```diff
@@ -1055,6 +1085,130 @@ static int get_user_host(const char *uh_line, unsigned int uh_len,
 #define S_ISDIR(x) ((x) & _S_IFDIR)
 #endif /*_WIN32 && !S_ISDIR*/
 
+/*
+  Flushes the buffer of log strings to the file.
+
+  @param all_messages Contains the log strings
+*/
+void flush_buffer(const std::vector<std::string>& all_messages) 
+{
+  std::string concatenated_messages;
+  if (logfile == NULL) return;
+  my_off_t filesize= logger_space_left(logfile); // High I/O called once every batch
+  my_off_t initial_filesize= file_rotate_size;
+
+  for (std::size_t i= 0; i < all_messages.size(); i++) 
+  {
+    if (concatenated_messages.size() > filesize) 
+    {
+      if (!(is_active= (logger_write_r(logfile,
+                                       1, 
+                                       concatenated_messages.c_str(), 
+                                       concatenated_messages.length()) 
+                        == 
+                        (int)concatenated_messages.length()))) 
+      {
+        ++log_write_failures;
+      }
+      concatenated_messages.clear();
+      filesize= initial_filesize; // Start with initial file size each time you reach 0
+    }
+    std::string msg= all_messages[i];
+    concatenated_messages= concatenated_messages + msg;
+    filesize-= msg.size(); // Decrement the filesize to keep track of if rotation is needed
+  }
+  if (concatenated_messages != "") 
+  {
+    if (!(is_active= (logger_write_r(logfile,
+                                       1, 
+                                       concatenated_messages.c_str(), 
+                                       concatenated_messages.length()) 
+                        == 
+                        (int)concatenated_messages.length())))
+    {
+      ++log_write_failures;
+    }
+  }
+}
+
+/*
+  Flushes and then resizes the queue
+*/
+static void resize_flush(size_t new_size) 
+{
+  std::unique_lock<std::mutex> lock(queue_mutex);
+  if (log_queue.size() != 0) 
+  {
+      flush_buffer(log_queue);
+  }
+  log_queue.clear();
+  max_size_log_queue= new_size;
+}
+
+/*
+  Call back function for the timer thread 
+*/
+void call_back_flush(void* arg) 
+{
+  std::unique_lock<std::mutex> lock(queue_mutex);
+  if (log_queue.size() != 0) 
+  {
+      flush_buffer(log_queue);
+      log_queue.clear();
+  }
+}
+
+
+void push_log(const std::string& message) 
+{
+  std::unique_lock<std::mutex> lock(queue_mutex);  
+  log_queue.emplace_back(message);
+  if (log_queue.size() >= max_size_log_queue)
+  {
+      flush_buffer(log_queue);
+      log_queue.clear();
+      return;
+  }
+  
+  if (log_queue.size() == 1 && audit_timer.expired) 
```

Comment:
I was wondering if you need to test this audit_timer.expired flag, then looked at the sources to find out that thr_timer_setting on non-expired timer is forbidden with a DBUG_ASSERT. I do not know the reasons behind that, but you could also do thr_timer_end (which does not destroy the timer itself, only removes from timer queue, if it is pending) prior to thr_timer_settime . This would be a tiny bit more thread-safe (all checks for "expired" flag under the same mutex). But I would not argue, even if you leave it as it is now.  Leaving it up to you.